File: index/chm/RecallTable.cpp
#include "RecallTable.hpp"

namespace chm {
	constexpr std::streamsize EF_SEARCH_WIDTH = 8;
	constexpr std::streamsize ELAPSED_PRETTY_WIDTH = 16;
	constexpr std::streamsize ELAPSED_WIDTH = 20;
	constexpr std::streamsize RECALL_WIDTH = 20;

	long long QueryBenchmark::getElapsedNum() const {
		return this->elapsed.count();
	}

	float QueryBenchmark::getRecall() const {
		return this->recall;
	}

	void QueryBenchmark::prettyPrintElapsed(std::ostream& s) const {
		prettyPrint(this->elapsed, s);
	}

	QueryBenchmark::QueryBenchmark(const uint efSearch) : efSearch(efSearch), elapsed(0), recall(0.f) {}

	void QueryBenchmark::setElapsed(const chr::microseconds& elapsed) {
		this->elapsed = elapsed;
	}

	void QueryBenchmark::setRecall(const float recall) {
		this->recall = recall;
	}

	RecallTable::RecallTable(const DatasetPtr& dataset, const std::vector<uint>& efSearchValues)
		: buildElapsed(0), dataset(dataset), efSearchValues(efSearchValues), index(nullptr) {

		this->benchmarks.reserve(this->efSearchValues.size());
	}

	RecallTable::RecallTable(const fs::path& datasetPath, const std::vector<uint>& efSearchValues)
		: RecallTable(std::make_shared<Dataset>(datasetPath), efSearchValues) {}

	void RecallTable::print(std::ostream& s) const {
		std::ios streamState(nullptr);
		streamState.copyfmt(s);

		this->dataset->writeShortDescription(s);
		s << this->index->getString() << '\n';
		s << "Build time: [";
		prettyPrint(this->buildElapsed, s);
		s << "], " << this->buildElapsed.count() << " us\n\n";

		s << std::setw(EF_SEARCH_WIDTH) << "EfSearch"
			<< std::setw(RECALL_WIDTH) << "Recall"
			<< std::setw(ELAPSED_PRETTY_WIDTH) << "Elapsed (pretty)"
			<< std::setw(ELAPSED_WIDTH) << "Elapsed (us)"
			<< std::setw(1) << '\n';

		for(const auto& benchmark : this->benchmarks) {
			s << std::setw(EF_SEARCH_WIDTH) << benchmark.efSearch << std::setw(RECALL_WIDTH);
			chm::print(benchmark.getRecall(), s, 3);
			s << std::setw(ELAPSED_PRETTY_WIDTH);
			benchmark.prettyPrintElapsed(s);
			s << std::setw(ELAPSED_WIDTH) << benchmark.getElapsedNum() << std::setw(1) << '\n';
		}

		s.copyfmt(streamState);
	}

	void RecallTable::run(const uint efConstruction, const uint mMax, const uint seed, std::ostream& s) {
		Timer timer{};
		this->benchmarks.clear();

		s << "Building index.\n";

		timer.reset();
		this->index = this->dataset->getIndex(efConstruction, mMax, seed);
		this->dataset->build(this->index);
		this->buildElapsed = timer.getElapsed();

		s << "Index built in ";
		prettyPrint(this->buildElapsed, s);
		s << "\n\n";

		for(const auto& efSearch : this->efSearchValues) {
			auto& benchmark = this->benchmarks.emplace_back(efSearch);

			s << "Querying with efSearch = " << efSearch << ".\n";

			timer.reset();
			const auto knnResults = this->dataset->query(this->index, efSearch);
			benchmark.setElapsed(timer.getElapsed());

			s << "Completed in ";
			benchmark.prettyPrintElapsed(s);
			s << "\nComputing recall.\n";

			timer.reset();
			benchmark.setRecall(this->dataset->getRecall(knnResults));
			const auto recallElapsed = timer.getElapsed();
			s << "Recall " << benchmark.getRecall() << " computed in ";
			prettyPrint(recallElapsed, s);
			s << "\n\n";
		}
	}

	chr::microseconds Timer::getElapsed() const {
		return chr::duration_cast<chr::microseconds>(chr::steady_clock::now() - this->start);
	}

	void Timer::reset() {
		this->start = chr::steady_clock::now();
	}

	Timer::Timer() {
		this->reset();
	}

	void prettyPrint(const chr::microseconds& elapsed, std::ostream& s) {
		chr::microseconds elapsedCopy = elapsed;
		const auto hours = convert<chr::hours>(elapsedCopy);
		const auto mins = convert<chr::minutes>(elapsedCopy);
		const auto secs = convert<chr::seconds>(elapsedCopy);
		const auto ms = convert<chr::milliseconds>(elapsedCopy);

		std::ios streamState(nullptr);
		streamState.copyfmt(s);

		print(hours, s);
		s << ':';
		print(mins, s);
		s << ':';
		print(secs, s);
		s << '.';
		print(ms, s);
		s << '.';
		print(elapsedCopy.count(), s);

		s.copyfmt(streamState);
	}

	void print(const float number, std::ostream& s, const std::streamsize places = 2) {
		std::ios streamState(nullptr);
		streamState.copyfmt(s);
		s << std::fixed << std::showpoint << std::setprecision(places) << number;
		s.copyfmt(streamState);
	}

	void print(const long long number, std::ostream& s, const std::streamsize places = 2) {
		s << std::setfill('0') << std::setw(places) << number;
	}
}

File: index/apps/helpers/RecallTable.py
from chm_hnsw import getRecall, Index, Space
from dataclasses import dataclass
from .Dataset import Dataset
import pandas
from pathlib import Path
import time

EF_SEARCH_WIDTH = 8
ELAPSED_PRETTY_WIDTH = 29
ELAPSED_WIDTH = 29
N = "\n"
RECALL_WIDTH = 12

def getPrettyElapsedStr(elapsedNS: int):
	return str(pandas.Timedelta(nanoseconds=elapsedNS))

@dataclass
class QueryBenchmark:
	efSearch: int
	elapsedNS: int = None
	recall: float = None

	def getPrettyElapsedStr(self):
		return getPrettyElapsedStr(self.elapsedNS)

class RecallTable:
	def __init__(self, dataset: Dataset, efSearchValues: list[int]):
		self.benchmarks: list[QueryBenchmark] = None
		self.buildElapsedNS: int = None
		self.dataset = dataset
		self.efSearchValues = efSearchValues
		self.index: Index = None
		self.space = Space.ANGULAR if self.dataset.angular else Space.EUCLIDEAN

	def __str__(self):
		return "\n".join([
			str(self.dataset),
			str(self.index),
			f"Build time: [{getPrettyElapsedStr(self.buildElapsedNS)}], {self.buildElapsedNS} ns{N}",
			f"{'EfSearch':>{EF_SEARCH_WIDTH}}{'Recall':>{RECALL_WIDTH}}{'Elapsed (pretty)':>{ELAPSED_PRETTY_WIDTH}}{'Elapsed (ns)':>{ELAPSED_WIDTH}}",
			*[
				f"{b.efSearch:>{EF_SEARCH_WIDTH}}{b.recall:>{RECALL_WIDTH}.3f}{b.getPrettyElapsedStr():>{ELAPSED_PRETTY_WIDTH}}{b.elapsedNS:>{ELAPSED_WIDTH}}"
				for b in self.benchmarks
			]
		])

	@classmethod
	def fromHDF(cls, datasetPath: Path, efSearchValues: list[int]):
		return cls(Dataset.fromHDF(datasetPath), efSearchValues)

	def run(self, efConstruction: int, mMax: int, seed: int):
		self.benchmarks = []
		print("Building index.")

		start = time.perf_counter_ns()
		self.index = Index(self.dataset.dim, efConstruction, self.dataset.trainCount, mMax, seed, self.space)
		self.index.push(self.dataset.train)
		end = time.perf_counter_ns()

		self.buildElapsedNS = end - start
		print(f"Index built in {getPrettyElapsedStr(self.buildElapsedNS)}.", end = "\n\n")

		for efSearch in self.efSearchValues:
			benchmark = QueryBenchmark(efSearch)
			self.benchmarks.append(benchmark)
			print(f"Querying with efSearch = {efSearch}.")

			start = time.perf_counter_ns()
			self.index.setEfSearch(efSearch)
			knnResults = self.index.query(self.dataset.test, self.dataset.k)
			end = time.perf_counter_ns()
			benchmark.elapsedNS = end - start
			print(f"Completed in {benchmark.getPrettyElapsedStr()}.")

			print("Computing recall.")
			start = time.perf_counter_ns()
			benchmark.recall = getRecall(self.dataset.neighbors, knnResults[0])
			end = time.perf_counter_ns()
			recallElapsedNS = end - start
			print(f"Recall computed in {getPrettyElapsedStr(recallElapsedNS)}.", end="\n\n")
